File: app/src/main/java/cz/civilizacehra/cipherbreaker/Azimuth.kt
package cz.civilizacehra.cipherbreaker

import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin

internal object Azimuth {
    private const val EARTH_RADIUS_METERS = 6371000

    /**
     * Returns latitude and longitude of the point reached from given position
     * after travelling given distance (in meters) under given azimuth (in degrees).
     */
    fun destination(lat: Double, lon: Double, distance: Double, angle: Double): Pair<Double, Double> {
        val distRatio = distance / EARTH_RADIUS_METERS
        val radAngle = toRad(angle)
        val radLat = toRad(lat)
        val radLon = toRad(lon)

        val newLat = asin(sin(radLat) * cos(distRatio) + cos(radLat) * sin(distRatio) * cos(radAngle))

        val newLon = radLon + atan2(sin(radAngle) * sin(distRatio) * cos(radLat),
                cos(distRatio) - sin(radLat) * sin(newLat))

        return Pair(toDeg(newLat), toDeg(newLon))
    }

    private fun toRad(degrees: Double): Double {
        return degrees * Math.PI / 180
    }

    private fun toDeg(radians: Double): Double {
        return radians * 180 / Math.PI
    }
}

File: pwa/src/logic/azimuth.ts
// Port of Azimuth.kt

const EARTH_RADIUS_METERS = 6371000;

const toRad = (degrees: number) => degrees * Math.PI / 180;
const toDeg = (radians: number) => radians * 180 / Math.PI;

/**
 * Returns latitude and longitude of the point reached from given position
 * after travelling given distance (in meters) under given azimuth (in degrees).
 */
export function destination(lat: number, lon: number, distance: number, angle: number): [number, number] {
  const distRatio = distance / EARTH_RADIUS_METERS;
  const radAngle = toRad(angle);
  const radLat = toRad(lat);
  const radLon = toRad(lon);

  const newLat = Math.asin(Math.sin(radLat) * Math.cos(distRatio) + Math.cos(radLat) * Math.sin(distRatio) * Math.cos(radAngle));
  const newLon = radLon + Math.atan2(Math.sin(radAngle) * Math.sin(distRatio) * Math.cos(radLat),
    Math.cos(distRatio) - Math.sin(radLat) * Math.sin(newLat));

  return [toDeg(newLat), toDeg(newLon)];
}
